Jenkins Failure: Polish Workchain Error

@sphuber I might need some help deciphering your RPN workchain 😬 

Basically, Jenkins is failing on this expression: "3 1 3 4 -4 2 * + + ^ ^", because it chooses to use the `add_calculation` function for [1, -2], when I guess it should choose `subtract_calcfunction`, or `settings.allow_negative = True` should be set:

<details>
<summary> 
verdi run /tmp/ci/polish/cli.py -X $CODE -C -F -t $TIMEOUT "3 1 3 4 -4 2 * + + ^ ^"
</summary> 
07/14/2020 10:51:58 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [149|Polish00WorkChain|setup]: Initial operands [3, 1, 3, 4, -4, 2]
07/14/2020 10:51:58 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [149|Polish00WorkChain|raise_power]: launching Polish01WorkChain<152> with operands 1 3 4 -4 2
07/14/2020 10:51:58 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [149|Polish00WorkChain|raise_power]: launching Polish01WorkChain<153> with operands 1 3 4 -4 2
07/14/2020 10:51:58 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [149|Polish00WorkChain|raise_power]: launching Polish01WorkChain<154> with operands 1 3 4 -4 2
07/14/2020 10:51:59 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [152|Polish01WorkChain|setup]: Initial operands [1, 3, 4, -4, 2]
07/14/2020 10:51:59 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [152|Polish01WorkChain|raise_power]: launching Polish02WorkChain<157> with operands 3 4 -4 2
07/14/2020 10:51:59 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [153|Polish01WorkChain|setup]: Initial operands [1, 3, 4, -4, 2]
07/14/2020 10:51:59 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [153|Polish01WorkChain|raise_power]: launching Polish02WorkChain<160> with operands 3 4 -4 2
07/14/2020 10:51:59 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [154|Polish01WorkChain|setup]: Initial operands [1, 3, 4, -4, 2]
07/14/2020 10:52:00 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [154|Polish01WorkChain|raise_power]: launching Polish02WorkChain<163> with operands 3 4 -4 2
07/14/2020 10:52:00 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [157|Polish02WorkChain|setup]: Initial operands [3, 4, -4, 2]
07/14/2020 10:52:00 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [157|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<166> with inputs uuid: bacaae37-6858-4872-8522-035ec8467968 (pk: 164) value: 0 and 3.0
07/14/2020 10:52:00 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [160|Polish02WorkChain|setup]: Initial operands [3, 4, -4, 2]
07/14/2020 10:52:00 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [160|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<169> with inputs uuid: d02c9b4d-cf25-46af-8e3b-d357206724aa (pk: 167) value: 0 and 3.0
07/14/2020 10:52:00 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [163|Polish02WorkChain|setup]: Initial operands [3, 4, -4, 2]
07/14/2020 10:52:01 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [163|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<172> with inputs uuid: 26171e88-12ea-455d-ac67-f3815f1255af (pk: 170) value: 0 and 3.0
07/14/2020 10:52:11 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [157|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<184> with inputs uuid: 5df1f9d5-3314-4e62-88dd-0c1256d2dcd9 (pk: 182) value: 3 and 4.0
07/14/2020 10:52:11 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [160|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<187> with inputs uuid: 755c1e69-f0dd-40c7-9712-b8ad65e35a76 (pk: 185) value: 3 and 4.0
07/14/2020 10:52:11 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [163|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<190> with inputs uuid: 8ac8dcb1-4f32-4214-b0e9-b40f8e932f5d (pk: 188) value: 3 and 4.0
07/14/2020 10:52:22 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [157|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<202> with inputs uuid: b6f9deae-d985-4dc3-b43f-11129da871d2 (pk: 200) value: 7 and -2
07/14/2020 10:52:22 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [160|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<205> with inputs uuid: f34a9c70-5ba5-4b6a-b4b1-5f74f799b97d (pk: 203) value: 7 and -2
07/14/2020 10:52:22 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [163|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<208> with inputs uuid: 3812aaac-7354-4849-baa5-86c509a667ca (pk: 206) value: 7 and -2
07/14/2020 10:52:33 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [157|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<220> with inputs uuid: 56d497be-8ea4-46c8-b99e-c46537e54149 (pk: 218) value: 5 and -2
07/14/2020 10:52:33 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [160|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<223> with inputs uuid: 49fefcfa-21ef-4b5f-9ffe-d581cbd5abd3 (pk: 221) value: 5 and -2
07/14/2020 10:52:33 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [163|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<226> with inputs uuid: 317d2940-6610-4718-ab78-dacf70faa578 (pk: 224) value: 5 and -2
07/14/2020 10:52:44 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [157|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<238> with inputs uuid: 3a8c8950-eae4-41a3-92da-f8c1daa46b16 (pk: 236) value: 3 and -2
07/14/2020 10:52:44 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [160|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<241> with inputs uuid: fbb7167e-3831-4bd8-88d2-38692e291d7b (pk: 239) value: 3 and -2
07/14/2020 10:52:45 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [163|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<244> with inputs uuid: cc0a95e9-d9e2-4068-b9bd-9d783383756b (pk: 242) value: 3 and -2
07/14/2020 10:52:55 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [157|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<256> with inputs uuid: 06cf52c0-3b75-49e8-a1ce-c8c4cfcb7188 (pk: 254) value: 1 and -2
07/14/2020 10:52:55 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [160|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<259> with inputs uuid: d1a1c9dd-df84-43d0-9d30-82dd2d7235f7 (pk: 257) value: 1 and -2
07/14/2020 10:52:56 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [163|Polish02WorkChain|add_calculation]: launching ArithmeticAddCalculation<262> with inputs uuid: c1d8e6d7-e063-4358-a49b-a97716ae0184 (pk: 260) value: 1 and -2
07/14/2020 10:53:07 AM <1937> aiida.parser.ArithmeticAddParser: [ERROR] parser returned exit code<410>: The sum of the operands is a negative number.
07/14/2020 10:53:07 AM <1937> aiida.parser.ArithmeticAddParser: [ERROR] parser returned exit code<410>: The sum of the operands is a negative number.
07/14/2020 10:53:07 AM <1937> aiida.parser.ArithmeticAddParser: [ERROR] parser returned exit code<410>: The sum of the operands is a negative number.
07/14/2020 10:53:08 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [157|Polish02WorkChain|post_add]: no output node found
07/14/2020 10:53:08 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [160|Polish02WorkChain|post_add]: no output node found
07/14/2020 10:53:08 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [163|Polish02WorkChain|post_add]: no output node found
07/14/2020 10:53:08 AM <1937> aiida.orm.nodes.process.workflow.workchain.WorkChainNode: [REPORT] [152|Polish01WorkChain|on_except]: Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/aiida/orm/utils/managers.py", line 83, in __getattr__
    return self._get_node_by_link_label(label=name)
  File "/opt/conda/lib/python3.7/site-packages/aiida/orm/utils/managers.py", line 64, in _get_node_by_link_label
    return self._node.get_outgoing(link_type=self._link_type).get_node_by_label(label)
  File "/opt/conda/lib/python3.7/site-packages/aiida/orm/utils/links.py", line 302, in get_node_by_label
    raise exceptions.NotExistent('no neighbor with the label {} found'.format(label))
aiida.common.exceptions.NotExistent: no neighbor with the label result found
</details>

I'm also guessing it has something to do with this commit https://github.com/aiidateam/aiida-core/commit/62a108ae3b30b9b096c9efdc8d80f26841649c12 (will keep digging)

I noted that in: https://github.com/aiidateam/aiida-core/blob/b316ab3ee6b512dbeefaa0aaa1368c4cdf6823ec/.ci/polish/lib/template/workchain.tpl#L48-L49
it only consider one of the operands to decide whether to use add or subtract, and in this case it is the 2nd operator that is negative. But then this appears to work fine for other expressions (e.g.` 1 -2 -1 4 -5 -5 * * * * +`)

Also perhaps of note is that the expression in question at one point has a negative power (if I've worked out RPN correctly!) which is generally prohibited:

```
3 1 3 4 -4 2 * + + ^ ^
-4 * 2 = -8
4 + -8 = -4
3 + -4 = -1
1 ^ -1 = 1
3 ^ 1 = 3
```

cc @giovannipizzi

--- a/aiida/parsers/plugins/arithmetic/add.py
+++ b/aiida/parsers/plugins/arithmetic/add.py
@@ -33,10 +33,10 @@ class ArithmeticAddParser(Parser):
         except ValueError:
             return self.exit_codes.ERROR_INVALID_OUTPUT
 
+        self.out('sum', Int(result))
+
         if result < 0:
             return self.exit_codes.ERROR_NEGATIVE_NUMBER
-
-        self.out('sum', Int(result))
 
 
 class SimpleArithmeticAddParser(Parser):

